python/ray/serve/_private/router.py

```diff
@@ -993,21 +989,37 @@ def __init__(
                     LongPollNamespace.RUNNING_REPLICAS,
                     deployment_id,
                 ): self._replica_scheduler.update_running_replicas,
+                (
+                    LongPollNamespace.AUTOSCALING_CONFIG,
+                    deployment_id,
+                ): self.update_autoscaling_config,
             },
             call_in_event_loop=event_loop,
         )
 
-        # Start the metrics pusher if autoscaling is enabled.
-        deployment_route = DeploymentRoute.FromString(
-            ray.get(controller_handle.get_deployment_info.remote(*deployment_id))
-        )
-        deployment_info = DeploymentInfo.from_proto(deployment_route.deployment_info)
         self.metrics_pusher = None
-        if deployment_info.deployment_config.autoscaling_config:
+        self.autoscaling_enabled = False
+        self.push_metrics_to_controller = controller_handle.record_handle_metrics.remote
+
+    def update_autoscaling_config(self, autoscaling_config):
+        self.autoscaling_config = autoscaling_config
+
+        # Start the metrics pusher if autoscaling is enabled.
+        if self.autoscaling_config:
             self.autoscaling_enabled = True
-            self.push_metrics_to_controller = (
-                controller_handle.record_handle_metrics.remote
-            )
+
+            # Optimization for autoscaling cold start time. If there are
+            # currently 0 replicas for the deployment, and there is at
+            # least one queued query on this router, then immediately
+            # push handle metric to the controller.
+            if (
+                len(self._replica_scheduler.curr_replicas) == 0
+                and self.num_queued_queries
+            ):
```

Comment:
We don't need to do this in this PR, but is there a reason we only push the metrics if there's a pending request? Is there any harm in pushing the metrics whenever the autoscaling config is updated?

Reply:
No, there is no harm. This optimization is specifically for cold start time though; e.g. if the metrics are pushed once every 10 seconds, and a request arrives at +1s, then the long poll client receives the config at +1.1s, we want to notify the controller immediately so that the time to first response isn't lower-bounded by 9s + [replica startup time].